patches/tock/01-persistent-storage.patch

```diff
@@ -415,7 +415,7 @@ index c52754be3..ae6a58341 100644
 +                .mpu()
 +                .allocate_region(
 +                    storage_location.address as *const u8,
-+                    storage_location.size,
++                    storage_location.unallocated_size,
```

Comment:
The reason I added `unallocated_size` is because using `size` doesn't work. There is already a `size` argument and we use `size` there. The `unallocated_size` argument is to tell the MPU that it's fine to allocate up to that size which is needed because the MPU is not able to allocate small regions near the end of the flash.

The `unallocated_size` is not used to prevent that part of the flash from being allocated by other things. This is just a workaround for the MPU. There is indeed no need to reserve that space.

Reply:
And we actually need to pass that number down as part of `StorageLocation`? We can't just round `size` up to the next power of 2?

Reply:
Right, we could assume the `unallocated_size` being the smallest number such that the allocation works:
- PRO: we don't specify the unallocated_size
- CON: we don't specify the unallocated_size so we don't know what region is actually used
- PRO: We don't touch the MPU code.
- CON: We change the MPU code.

I'm fine to try that if we think it's better.

Reply:
Why touch MPU code? It would be replacing `unallocated_size` here in `kernel/src/process.rs` with the rounded up `size`?
The argument of not knowing the region used is technically true for the app too, since the MPU rounds up its region size too. And the actual app size is even more implicit (even nondeterministic to some small degree?).

I like the cleaner board definition. Maybe more freedom to assign regions will be required for some boards, but we can always increase complexity if necessary?

Reply:
> Why touch MPU code? It would be replacing unallocated_size here in kernel/src/process.rs with the rounded up size?

The logic is a bit complicated and I'm not sure rounding up always works. But I'll try and if it works let's do that. This is just a workaround for Nordic and we probably won't change those values often.

> The argument of not knowing the region used is technically true for the app too, since the MPU rounds up its region size too. And the actual app size is even more implicit (even nondeterministic to some small degree?).

But it doesn't do it outside of the flash already use by the binary. This is why elf2tab already does the rounding up. In our case, we would round outside the requested size. (Note that it's probably possible to improve the MPU code, quickly looking through it, it seems to miss the opportunity to use a smaller subregion size when possible.)

> I like the cleaner board definition. Maybe more freedom to assign regions will be required for some boards, but we can always increase complexity if necessary?

￼Yes, I'm fine with the hack in process.rs to use some heuristics to make it work for Nordic. We'll need to test it on other boards too.

Reply:
We might need a better MPU implementation down the road when we want more precise control over regions in the future, i.e. on boards with less flash. So hack for Nordic might be what we want for now. Sounds good.